File: 12-Trees/02-BST/java/ValidateBST.java
public class ValidateBST {
    static class TreeNode { int val; TreeNode left, right; TreeNode(int x) { val = x; } }
    static boolean isValid(TreeNode node, long min, long max) {
        if (node == null) return true;
        if (node.val <= min || node.val >= max) return false;
        return isValid(node.left, min, node.val) && isValid(node.right, node.val, max);
    }
    public static boolean isValidBST(TreeNode root) { return isValid(root, Long.MIN_VALUE, Long.MAX_VALUE); }
    public static void main(String[] args) {
        TreeNode root = new TreeNode(2); root.left = new TreeNode(1); root.right = new TreeNode(3);
        System.out.println(isValidBST(root)); // true
    }
}

File: 12-Trees/02-BST/cpp/validate_bst.cpp
/*
 * Problem: Validate Binary Search Tree
 * Source:  LeetCode 98
 * Time: O(n), Space: O(h)
 * Tags: bst, tree, recursion
 */
#include <bits/stdc++.h>
using namespace std;
struct TreeNode { int val; TreeNode *left, *right; TreeNode(int x):val(x),left(nullptr),right(nullptr){} };

bool isValid(TreeNode* node, long minVal, long maxVal) {
    if (!node) return true;
    if (node->val <= minVal || node->val >= maxVal) return false;
    return isValid(node->left, minVal, node->val) && isValid(node->right, node->val, maxVal);
}

bool isValidBST(TreeNode* root) { return isValid(root, LONG_MIN, LONG_MAX); }

int main() {
    TreeNode* root = new TreeNode(5);
    root->left = new TreeNode(1); root->right = new TreeNode(4);
    root->right->left = new TreeNode(3); root->right->right = new TreeNode(6);
    cout << boolalpha << isValidBST(root) << endl; // false
    return 0;
}
